tutorials/developer_api_guide/my_trainable_tensor_subclass.py

```diff
@@ -0,0 +1,200 @@
+
+implements = MyTrainableDTypeTensor.implements
+
+class _QuantizedLinearOp(torch.autograd.Function):
+    """
+    Forward and backward definition for linear with quantized weights.
+    Weights are dequantized during both the forward and the backward passes.
+    """
+
+    @staticmethod
+    def forward(
+        ctx: torch.autograd.function.FunctionCtx,
+        input_tensor: torch.Tensor,
+        weight_tensor: torch.Tensor,
+    ) -> torch.Tensor:
+        assert isinstance(weight_tensor, MyTrainableDTypeTensor)
+        ctx.save_for_backward(input_tensor, weight_tensor)
+        weight_tensor = weight_tensor.dequantize()
+        return torch.nn.functional.linear(input_tensor, weight_tensor)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input_tensor, weight_tensor = ctx.saved_tensors
+        weight_int = weight_tensor.layout_tensor.int_data
+        weight_scale = weight_tensor.layout_tensor.scale
+        grad_input = (grad_output * weight_scale) @ weight_int.to(grad_output.dtype)
+        grad_weight = grad_output.T @ input_tensor
+        return grad_input, grad_weight
+
+@implements(torch.nn.functional.linear)
+def _(func, types, args, kwargs):
+    """
+    Handle the linear op with quantized weights.
+    For simplicity, we run both the forward and backward passes entirely in float.
+    """
+    assert isinstance(args[1], MyTrainableDTypeTensor)
+    if len(args) > 2 and args[2] is not None:
+        raise NotImplementedError("linear bias not yet supported")
+    return _QuantizedLinearOp.apply(args[0], args[1])
+
+@implements(aten.add_.Tensor)
+def _(func, types, args, kwargs):
+    """
+    Handle the in-place add op, called by the optimizer to update
+    the quantized weight during training.
+    """
+    assert len(args) == 2
+    assert isinstance(args[0], MyTrainableDTypeTensor)
+    assert args[0].layout_tensor.int_data.dtype == torch.int8
+    float0 = args[0].dequantize()
+    float1 = args[1].dequantize() if isinstance(args[1], MyTrainableDTypeTensor) else args[1]
+    new_value = torch.add(float0, float1, **kwargs)
+    new_layout_tensor = MyTrainableDTypeTensor._quantize(
+        new_value,
+        args[0].layout_tensor.get_layout_type(),
+    )
+    args[0].layout_tensor = new_layout_tensor
+    return return_and_correct_aliasing(func, args, kwargs, args[0])
+
+@implements(aten.add.Tensor)
+def _(func, types, args, kwargs):
+    """Handle the add op, called by the optimizer during training."""
+    assert len(args) == 2
+    assert not isinstance(args[0], MyTrainableDTypeTensor)
+    assert isinstance(args[1], MyTrainableDTypeTensor)
+    out = torch.add(args[0], args[1].dequantize(), **kwargs)
+    return return_and_correct_aliasing(func, args, kwargs, out)
+
+
+########
+# Test #
+########
+
+class M(torch.nn.Module):
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.linear = torch.nn.Linear(1024, 1024, bias=False)
```

Comment:
nit: maybe changing to things like 512, 1024 so we can catch shape problems as well